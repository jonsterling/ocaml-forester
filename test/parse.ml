open Forester.Parse
open Core

let emit _ = () (* ignore *)

let fatal _ = exit 1

let _ =
  Core.Reporter.run ~emit ~fatal @@ fun () ->
  let good = Result.get_ok @@ parse_string {|
    \title{Good}
    \taxon{Test}
    \author{Testy}

    \p{
      This should parse correctly.
    }
    |}
  in
  Printf.printf "parse_good_result:\n%s\n\n" (Code.show good)

let _ =
  Core.Reporter.run ~emit ~fatal @@ fun () ->
  let bad, errors = Result.get_error @@ parse_string {|
    \title{Error recovery}
    \taxon{Test}
    \author{Testy}

    \p{
      This has some syntax errors, but the parser should recover as much as possible.
    }

    Invalid naked text here.

    \p{
      More good text.
    }

    \p{
      Some bad text [
    }

    \p{
      Even more good text.
    }

    \p{
      Some more bad ] text
    }

    \p{
      Skipped an ending brace here.
    |}
  in
  errors |> List.iter (fun (e : Core.Reporter.Message.t Asai.Diagnostic.t) -> Printf.printf "error: %s\n" (Asai.Diagnostic.string_of_text e.explanation.value));
  Printf.printf "parse_bad_result:\n%s\n\n" (Code.show bad)
