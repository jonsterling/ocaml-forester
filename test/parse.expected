parse_good_result:
[(Code.Title [(Code.Text "Good")]); (Code.Taxon "Test");
  (Code.Author "Testy");
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "This");
       (Code.Text " "); (Code.Text "should"); (Code.Text " ");
       (Code.Text "parse"); (Code.Text " "); (Code.Text "correctly.");
       (Code.Text "\n"); (Code.Text "    ")]
     ))
  ]

error: syntax error, unexpected `Invalid` 
error: syntax error, unexpected `naked` 
error: syntax error, unexpected `text` 
error: syntax error, unexpected `here.` 
error: syntax error, unexpected `}` 
error: syntax error, unexpected `]` 
error: syntax error, unexpected `` 
parse_bad_result:
[(Code.Title [(Code.Text "Error"); (Code.Text " "); (Code.Text "recovery")]);
  (Code.Taxon "Test"); (Code.Author "Testy");
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "This");
       (Code.Text " "); (Code.Text "has"); (Code.Text " ");
       (Code.Text "some"); (Code.Text " "); (Code.Text "syntax");
       (Code.Text " "); (Code.Text "errors,"); (Code.Text " ");
       (Code.Text "but"); (Code.Text " "); (Code.Text "the");
       (Code.Text " "); (Code.Text "parser"); (Code.Text " ");
       (Code.Text "should"); (Code.Text " "); (Code.Text "recover");
       (Code.Text " "); (Code.Text "as"); (Code.Text " ");
       (Code.Text "much"); (Code.Text " "); (Code.Text "as");
       (Code.Text " "); (Code.Text "possible."); (Code.Text "\n");
       (Code.Text "    ")]
     ));
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "More");
       (Code.Text " "); (Code.Text "good"); (Code.Text " ");
       (Code.Text "text."); (Code.Text "\n"); (Code.Text "    ")]
     ));
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "Some");
       (Code.Text " "); (Code.Text "bad"); (Code.Text " ");
       (Code.Text "text"); (Code.Text " ");
       (Code.Group (Base.Squares, [(Code.Text "\n"); (Code.Text "    ")]))]
     ));
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "Even");
       (Code.Text " "); (Code.Text "more"); (Code.Text " ");
       (Code.Text "good"); (Code.Text " "); (Code.Text "text.");
       (Code.Text "\n"); (Code.Text "    ")]
     ));
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "Some");
       (Code.Text " "); (Code.Text "more"); (Code.Text " ");
       (Code.Text "bad"); (Code.Text " "); (Code.Text " ");
       (Code.Text "text"); (Code.Text "\n"); (Code.Text "    ")]
     ));
  (Code.Prim (`P,
     [(Code.Text "\n"); (Code.Text "      "); (Code.Text "Skipped");
       (Code.Text " "); (Code.Text "an"); (Code.Text " ");
       (Code.Text "ending"); (Code.Text " "); (Code.Text "brace");
       (Code.Text " "); (Code.Text "here."); (Code.Text "\n");
       (Code.Text "    ")]
     ))
  ]

